### ISSUE TITLE
Asteroid Belts

### ISSUE THREAD

**Issue by lumi-a** [Tim2othy/spacegame#32]
It should be possible to calculate stable orbits (we need only initial starting-point and initial velocity, but we'll likely derive those from ellipse-formulas) for objects of a given mass, and thus create asteroid-belts around some planets.

**Comment by Tim2othy** [comment]
Done!

There are various levels of details one can add.

1. The simplest implementation only has circular orbits, and one of the coordinates of the starting point of the asteroid is the same as the planets. I.e. it's only offset along one dimension. 
2. Next level allows for elliptical orbits but the asteroid always starts at the apoapsis, this is still simple to implement, now you need something called the `total_specific_energy` but it's still not too complicated. 
3. This level allows the offset to be different from zero along both dimensions.
4. The final level allows the asteroid to start anywhere during its orbit, i.e. apoapsis, periapsis or anywhere in-between. This is a lot more complicated; you need to use `eccentricity` and the `true_anomaly` and a lot more maths than before, but of course it's still not that complicated.

right now, the asteroid generation is implemented as a standalone function `generate_asteroid` that acts on a planet and the list of all asteroids. It should be simple to implement it as a method of the universe class.

**Comment by Tim2othy** [comment]
The one problem that remains...

If there are multiple planets the orbits aren't stable and the asteroids will crash or veer off into space. And if there are a lot of large asteroids, they will crash into each other and knock each other out of orbit.

**Comment by Tim2othy** [comment]
Asteroids are now more organized. The function that generates them is in the universe class now and, main doesn't need to import the Asteroid class anymore. 

The problem that they don't have stable orbits if there are multiple planets still remains. One way to solve this would be to have only around 4 Planets and create 50 asteroids per planet have the game quickly simulate their orbits for a while, and only then add the players by that time there might be only 5 asteroids per planet left but they probably have somewhat stable orbits.

A simple way to do this is to just set the gravitational constant to 500 or so for one second, then set it back to normal and add in the player.

This seems a bit ad hoc but I guess it's what the real universe also does. A bunch of asteroids are created some are unstable and crash into planets and some other ones randomly find stable orbits and chill there.

**Comment by lumi-a** [comment]
> One way to solve this would be to have only around 4 Planets

Why limit the number of planets? To save on the amount of computation that needs to be done up-front?

**Comment by Tim2othy** [comment]
> > One way to solve this would be to have only around 4 Planets
> 
> Why limit the number of planets? To save on the amount of computation that needs to be done up-front?

I think with more planets almost all asteroids would quickly crash into a planet, but this might not be true if the distances are large enough.

**Comment by Tim2othy** [comment]
I don't like the `OrbitAsteroid` solution.

**Comment by lumi-a** [comment]
We had the "cut gravity to 0 if it's small enough" idea in #11 already. Once implemented, we can close #11 and this issue.

**Comment by Tim2othy** [comment]
Yes, we should implement that, I'd guess that cutting gravity to 0 would only fix the issue if planets are quite far away and I don't think that would be too fun gameplay wise (different ships would just quickly fly past each other and bullets would always miss), but it probably improves the orbit stability somewhat.

**Comment by Tim2othy** [comment]
I'd close this for now, I think planets are mostly cooler than asteroids. We could also add moons that orbit planets (In our solar system there are two moons larger than mercury). Maybe once we have planets working well, and moons orbiting some of the larger planets we can add asteroids back in again if we want but for now, I'd consider this finished.

**Comment by lumi-a** [comment]
Does this mean we can remove Universe.generate_planet?

**Comment by Tim2othy** [comment]
> Does this mean we can remove Universe.generate_planet?

Hmm? Why would it? That's the way planets are added to the game.
What I meant was:
- Initially we wanted to add asteroids and asteroid belts
- Now, instead of asteroids orbiting Planets we have Planets orbiting a star
- So, this issue can be closed. In a sense we could mark this as "not planned" because asteroids don't exist anymore but in a different sense this issue was completed, except that asteroids are now planets and work a bit differently.

**Comment by lumi-a** [comment]
I see, I thought the method would now be responsible for creating moons around plants instead of planets around a star, but yes, in the latter case we can keep it.
